src/audio/copier/copier.c

```diff
@@ -780,6 +826,178 @@ static int copier_comp_trigger(struct comp_dev *dev, int cmd)
+	int16_t *w_ptr = audio_stream_wrap(dst, (int16_t *)dst->w_ptr + dst_channel);
+	int i;
+
+	for (i = 0; i < frame_count; i++) {
+		*w_ptr = *r_ptr;
+		r_ptr = audio_stream_wrap(src, r_ptr + src->channels);
+		w_ptr = audio_stream_wrap(dst, w_ptr + dst->channels);
+	}
+}
+
+static inline void copy_single_channel_c32(struct audio_stream __sparse_cache *dst,
+					   int dst_channel,
+					   const struct audio_stream __sparse_cache *src,
+					   int src_channel, int frame_count)
+{
+	int32_t *r_ptr = audio_stream_wrap(src, (int32_t *)src->r_ptr + src_channel);
+	int32_t *w_ptr = audio_stream_wrap(dst, (int32_t *)dst->w_ptr + dst_channel);
+	int i;
+
+	for (i = 0; i < frame_count; i++) {
+		*w_ptr = *r_ptr;
+		r_ptr = audio_stream_wrap(src, r_ptr + src->channels);
+		w_ptr = audio_stream_wrap(dst, w_ptr + dst->channels);
+	}
+}
+
+static void demux_from_multi_endpoint_buffer(struct copier_data *cd)
+{
+	struct comp_buffer __sparse_cache *multi_buf_c, *endp_buf_c;
+	int endp_idx, endp_channel;
+	uint32_t avail_frames, avail_bytes, container_size;
+
+	multi_buf_c = buffer_acquire(cd->multi_endpoint_buffer);
+
+	/* NOTE: free space in destination buffers is not checked. This is done to have
+	 * behaviour consistent with copier_copy() which does not check for free space either.
+	 * However, quite probably, such behaviour is just a bug and have to be fixed in both
+	 * copier_copy() and here.
+	 */
+
+	avail_frames = audio_stream_get_avail_frames(&multi_buf_c->stream);
+	container_size = audio_stream_sample_bytes(&multi_buf_c->stream);
+
+	avail_bytes = avail_frames * audio_stream_frame_bytes(&multi_buf_c->stream);
+	buffer_stream_invalidate(multi_buf_c, avail_bytes);
+
+	for (endp_idx = 0; endp_idx < cd->endpoint_num; endp_idx++) {
+		uint32_t bytes_produced;
+
+		endp_buf_c = buffer_acquire(cd->endpoint_buffer[endp_idx]);
+
+		for (endp_channel = 0; endp_channel < endp_buf_c->stream.channels;
+				endp_channel++) {
+			int multi_buf_channel = endp_buf_c->chmap[endp_channel];
+
+			if (container_size == 2) {
+				copy_single_channel_c16(&endp_buf_c->stream, endp_channel,
+							&multi_buf_c->stream, multi_buf_channel,
+							avail_frames);
+			} else {
+				assert(container_size == 4);
+				copy_single_channel_c32(&endp_buf_c->stream, endp_channel,
+							&multi_buf_c->stream, multi_buf_channel,
+							avail_frames);
+			}
+		}
+
+		bytes_produced = avail_frames * audio_stream_frame_bytes(&endp_buf_c->stream);
+		buffer_stream_writeback(endp_buf_c, bytes_produced);
+		comp_update_buffer_produce(endp_buf_c, bytes_produced);
+		buffer_release(endp_buf_c);
+	}
+
+	comp_update_buffer_consume(multi_buf_c, avail_bytes);
+	buffer_release(multi_buf_c);
+}
+
+static void mux_into_multi_endpoint_buffer(struct copier_data *cd)
+{
+	struct comp_buffer __sparse_cache *multi_buf_c, *endp_buf_c;
+	int endp_idx, endp_channel;
+	uint32_t avail_frames = UINT32_MAX, container_size, bytes_produced;
```

Comment:
don't put on the same lines variables that you initialize and other that you don't initialize. This is painful to read and review.
I personally err on the side of one variable per line. Some coding standards also require this.